crypto3/libs/multiprecision/include/nil/crypto3/multiprecision/ops/import_export.hpp

```diff
@@ -0,0 +1,210 @@
+            *val.limbs() |= value;
+
+            //
+            // Check for overflow bits:
+            //
+            bit_location = limb_bits - bit_location;
+
+            NIL_CO3_MP_ASSERT(
+                !((bit_location < sizeof(bits) * CHAR_BIT) && (bits >>= bit_location)));
+        }
+
+        template<std::size_t Bits>
+        std::uintmax_t extract_bits(const big_uint<Bits>& val, std::size_t location,
+                                    std::size_t count) {
+            std::size_t limb = location / limb_bits;
+            std::size_t shift = location % limb_bits;
+            std::uintmax_t result = 0;
+            std::uintmax_t mask = count == std::numeric_limits<std::uintmax_t>::digits
+                                      ? ~static_cast<std::uintmax_t>(0)
+                                      : (static_cast<std::uintmax_t>(1u) << count) - 1;
+            if (count > (limb_bits - shift)) {
+                result = extract_bits(val, location + limb_bits - shift,
+                                      count - limb_bits + shift);
+                result <<= limb_bits - shift;
+            }
+            if (limb < val.limbs_count()) {
+                result |= (val.limbs()[limb] >> shift) & mask;
+            }
+            return result;
+        }
+
+        template<std::size_t Bits, typename Iterator>
+        big_uint<Bits>& import_bits_generic(big_uint<Bits>& result, Iterator i, Iterator j,
+                                            std::size_t chunk_size = 0, bool msv_first = true) {
+            big_uint<Bits> newval;
+
+            using value_type = typename std::iterator_traits<Iterator>::value_type;
+            using difference_type = typename std::iterator_traits<Iterator>::difference_type;
+            using size_type = typename std::make_unsigned<difference_type>::type;
+
+            if (!chunk_size) {
+                chunk_size = std::numeric_limits<value_type>::digits;
+            }
+
+            size_type limbs = std::distance(i, j);
+            size_type bits = limbs * chunk_size;
+
+            // TODO(ioxid): this breaks marshalling tests
+            // NIL_CO3_MP_ASSERT(bits <= Bits);
+
+            difference_type bit_location = msv_first ? bits - chunk_size : 0;
+            difference_type bit_location_change =
+                msv_first ? -static_cast<difference_type>(chunk_size) : chunk_size;
+
+            while (i != j) {
+                assign_bits(newval, *i, static_cast<std::size_t>(bit_location), chunk_size);
+                ++i;
+                bit_location += bit_location_change;
+            }
+
+            // This will remove the upper bits using upper_limb_mask.
+            newval.normalize();
+
+            result = std::move(newval);
+            return result;
+        }
+
+        template<std::size_t Bits, typename T>
+        inline big_uint<Bits>& import_bits_fast(big_uint<Bits>& result, T* i, T* j,
```

Comment:
Maybe move this, and other functions that must be used only from this file to an unnamed namespace, such that people won't call them from the outside. Kind of make them "private".

Reply:
In headers this wont actually work since they will be available for [qualified name lookup](https://en.cppreference.com/w/cpp/language/qualified_lookup) in parent namespace. I think that detail namespace better protects them from usage.